### ISSUE TITLE
Initialization of Embedding/NormalizedEmbedding

### ISSUE THREAD

**Issue by danpovey** [speechbrain/speechbrain#926]
I noticed something that might be a problem.

You are using NormalizedEmbedding which multiplies the embedding output by sqrt(dim).  This is good, because if you don't do this, the elements of the embedding parameter matrix will stay with a standard deviation around 1.0, which is way larger than other parameters and makes the embedding parameters train (relatively) much slower.  With the NormalizedEmbedding, the elements of that parameter matrix should approach very roughly 1.0/sqrt(dim) as training proceeds.  However, you seem to be using the default initialization of torch.nn.Embedding() which initializes the parameters to stddev=1.0.  This probably makes the job of initially training the parameters harder than it needs to be.

I just discovered a related problem in our own codebase, which is different, closer to ESPNet -- that the decoder Embedding parameters were training much slower than others.  E.g. from epoch 2->3 of full Librispeech, most parameters (norm of difference) / (norm of epoch-3 parameter) was in the range [0.5,1], but the decoder embedding was something like 0.08.     @sw005320 you might want to look at this issue as well.

**Comment by mravanelli** [comment]
Interesting observation. What do you think @popcornell? @johntremblay is refactoring the transformer part these days. He might wanna try an experiment to evaluate the difference.

**Comment by popcornell** [comment]
Sorry for the super late response. 

Thanks @danpovey for spotting this. I think it makes sense to change the initialization. 
I have run two experiments actually to see if in practice a difference is observed. 

I added after https://github.com/speechbrain/speechbrain/blob/b408b0c1062277370bc4efa2f2b2c9a79aebcc50/speechbrain/nnet/embedding.py#L101

`torch.nn.init.normal_(self.Embedding.weight, 0.0, (1/self.num_embeddings**0.5))` 

Here are the results: 

Standard: best-valid-wer:  3.05 |  test-clean: 2.25 | test-other 5.32
Normalized:  best-valid-wer:  3.12 | test-clean: 2.23 |test-other 5.26

standard init seems more prone to overfitting.

**Comment by danpovey** [comment]
OK-- I'd put those differences down to noise myself.
For us also, fixing this did not make much difference (although it did allow the embedding to train more meaningfully).
I suspect the reason is that they are only used as an input, so the random initialization works OK (the network can
work around it, whatever it is).

**Comment by Adel-Moumen** [comment]
Hello,

Sorry to bring this issue out of the abyss, but is there anything we should be doing about this issue? 

I don't see any reason to keep it alive.

Feel free to let me know what you think. 🙂

**Comment by Adel-Moumen** [comment]
Hello,

It seems there has been no activity for a long time, so I am closing this issue.

Feel free to reopen it if you still encounter the problem. Thank you. 🙂
